File: sc_custom/public/js/update_items.js
// "Tools > Update Item Names" on draft Material Request, Request for Quotation,
// Purchase Order and Stock Entry (FEAT-186). Mirrors BOM "Update Cost" (save=false):
// the server computes the changes (sc_custom/api/update_items.py), they are written
// into the open form and the user saves. Each doctype script calls
// sc_custom.update_items.add_button(frm) from its refresh handler.
frappe.provide('sc_custom.update_items');

sc_custom.update_items.add_button = function(frm) {
    if (frm.doc.docstatus !== 0 || !(frm.doc.items || []).length) return;
    frm.add_custom_button(__('Update Item Names'), function() {
        sc_custom.update_items.run(frm);
    }, __('Tools'));
};

sc_custom.update_items.run = function(frm) {
    const items = (frm.doc.items || []).map(row => ({
        name: row.name,
        idx: row.idx,
        item_code: row.item_code,
        item_name: row.item_name,
    }));

    frappe.call({
        method: 'sc_custom.api.update_items.get_item_updates',
        args: { doctype: frm.doc.doctype, items: items },
        freeze: true,
        callback(r) {
            const changes = r.message || [];
            if (!changes.length) {
                frappe.show_alert({ message: __('No changes in item names found'), indicator: 'blue' });
                return;
            }

            const row_doctype = frm.fields_dict.items.grid.doctype;
            changes.forEach(c => frappe.model.set_value(row_doctype, c.name, c.fieldname, c.new));
            frm.refresh_field('items');

            const rows = changes.map(c =>
                `<tr><td>${c.idx}</td><td>${frappe.utils.escape_html(c.item_code)}</td>` +
                `<td>${frappe.utils.escape_html(c.old)}</td><td>${frappe.utils.escape_html(c.new)}</td></tr>`
            ).join('');
            frappe.msgprint({
                title: __('Item Names Updated'),
                indicator: 'green',
                message: `<p>${__('{0} row(s) updated. Save the document to keep the changes.', [changes.length])}</p>` +
                    `<table class="table table-bordered table-sm"><thead><tr>` +
                    `<th>#</th><th>${__('Item Code')}</th><th>${__('Old Name')}</th><th>${__('New Name')}</th>` +
                    `</tr></thead><tbody>${rows}</tbody></table>`,
            });
        },
    });
};

File: sc_custom/api/update_items.py
"""Update Item Names — refresh the item name of the item rows from the Item master.

Used by the "Tools > Update Item Names" button on draft Material Request, Request for
Quotation, Purchase Order and Stock Entry (FEAT-186). The rows copy ``item_name`` once,
when the item code is picked, so a later rename of the Item never reaches existing
drafts.

Follows the flow of BOM "Update Cost" with ``save=false``: the server only computes the
changes, the client writes them into the open form and the user saves. Nothing is
written to the database here, so the button also works on unsaved documents.

``REFRESH_FIELDS`` maps Item master field -> row field. Only ``item_name`` for now;
``description`` is deliberately excluded because it is often edited by hand on the row.
"""

import json

import frappe
from frappe import _

SUPPORTED_DOCTYPES = ("Material Request", "Request for Quotation", "Purchase Order", "Stock Entry")
REFRESH_FIELDS = {"item_name": "item_name"}


@frappe.whitelist()
def get_item_updates(doctype, items):
	"""Return the row values that differ from the Item master.

	``items``: list of ``{name, idx, item_code, <row fields>}`` taken from the open form.
	Result: list of ``{name, idx, item_code, fieldname, old, new}``, one per changed value.
	"""
	if doctype not in SUPPORTED_DOCTYPES:
		frappe.throw(_("Update Item Names is not available for {0}").format(_(doctype)))
	frappe.has_permission("Item", "read", throw=True)

	if isinstance(items, str):
		items = json.loads(items)

	item_codes = {row.get("item_code") for row in items if row.get("item_code")}
	if not item_codes:
		return []

	master = {
		d.name: d
		for d in frappe.get_all(
			"Item",
			filters={"name": ("in", list(item_codes))},
			fields=["name", *REFRESH_FIELDS.keys()],
		)
	}

	changes = []
	for row in items:
		item = master.get(row.get("item_code"))
		if not item:
			continue
		for item_field, row_field in REFRESH_FIELDS.items():
			new_value = item.get(item_field) or ""
			old_value = row.get(row_field) or ""
			if new_value != old_value:
				changes.append(
					{
						"name": row.get("name"),
						"idx": row.get("idx"),
						"item_code": row.get("item_code"),
						"fieldname": row_field,
						"old": old_value,
						"new": new_value,
					}
				)
	return changes
